Translate Hooks FAQ (#123)
reactjs/id.reactjs.org: content/docs/hooks-faq.md

permalink: docs/hooks-faq.html
prev: hooks-reference.html
---
<!-- docs/hooks-faq.md ini diterjemahkan oleh dummyeuy (Muhammad Ghalib) -->
*Hooks* adalah tambahan baru dalam React 16.8. React memungkikan anda untuk menggunakan *state* dan fitur-fitur React lainnya tanpa perlu menuliskan sebuah kelas. 

Laman ini menjawab beberapa pertanyaan yang sering diajukan (*frequently asked questions*) tentang [Hooks](/docs/hooks-overview.html).

<!--
  jika Anda perlu me-egenerate ini, snippet dalam konsol devtools ini bisa membantu:

  $$('.anchor').map(a =>
    `${' '.repeat(2 * +a.parentNode.nodeName.slice(1))}` +
    `[${a.parentNode.textContent}](${a.getAttribute('href')})`
  ).join('\n')
-->

* **[Strategi Adopsi](#adoption-strategy)**
  * [Versi React mana yang termasuk *Hooks* di dalamnya?](#which-versions-of-react-include-hooks)

>   * [Apakah perlu menulis ulang semua kelas komponen saya?](#do-i-need-to-rewrite-all-my-class-components)

Author: Understood and (suggested) revision's applied

>   * [Apa yang dapat saya lakukan dengn *Hooks* namun tidak bisa saya lakukan dengan kelas-kelas?](#what-can-i-do-with-hooks-that-i-couldnt-with-classes)

Author: > [Apa yang dapat saya lakukan dengn *Hooks* namun tidak bisa saya lakukan dengan kelas?]

Suggested revision's applied

  * [Bagaimana cara *memoize* perhitungan?](#how-to-memoize-calculations)
  * [Bagaimana cara membuat objek *expensive* secara *lazy*?](#how-to-create-expensive-objects-lazily)
  * [Apakah *Hooks* lamban karena membuat fungsi di saat *render*?](#are-hooks-slow-because-of-creating-functions-in-render)
  * [Bagaimana cara menghindari pengoperan *callback* menjadi *down*?](#how-to-avoid-passing-callbacks-down)
  * [Bagimana cara membaca sebuah nilai yang sering berubah dari *useCallback*?](#how-to-read-an-often-changing-value-from-usecallback)
* **[*Under the Hood*](#under-the-hood)**
  * [Bagaimana React menghubungkan panggilan *Hook* dengan komponen?](#how-does-react-associate-hook-calls-with-components)
  * [Bagaimana *seni awal*-nya *Hooks*?](#what-is-the-prior-art-for-hooks)

## Strategi Adopsi {#adoption-strategy}

### Versi React mana yang termasuk *Hooks* di dalamnya? {#which-versions-of-react-include-hooks}

Mulai dari 16.8.0, dalam React sudah termasuk *stable implementation* dari React *Hooks* untuk:

* React *DOM*
* React *Native*
* React *DOM Server*
* React *Test Renderer*

> * React *Shallow Renderer*

Author: Understood.
Revision's applied

Perhatikan bahwa **untuk menjalankan Hooks, semua *package* React perlu setidaknya pada versi 16.8.0 atau lebih tinggi**. *Hooks* tidak akan bekerja jika Anda lupa melakukan pembaruan, sebagai contohnya, React *DOM*.

[React Native 0.59](https://facebook.github.io/react-native/blog/2019/03/12/releasing-react-native-059) dan versi di atasnya sudah mendukung Hooks.

### Apakah perlu menulis ulang semua kelas komponen saya?{#do-i-need-to-rewrite-all-my-class-components}

> Tidak. [Tidak ada rencana](/docs/hooks-intro.html#gradual-adoption-strategy) untuk menghilangkan kelas dari React -- kami perlu terus mengirimkan produk dan tidak akan melakukan penulisan ulang. Kami rekomendasikan untuk mencoba *Hooks* dalam kode baru.

Author: > Tidak. [Tidak ada rencana](/docs/hooks-intro.html#gradual-adoption-strategy) untuk menghilangkan kelas dari React -- kita semua perlu mengirimkan produk baru dan tidak memiliki waktu untuk penulisan ulang. Kami rekomendasikan untuk mencoba *Hooks* dalam kode baru.

Understood and (suggested) revision's applied

> ### Apa yang dapat saya lakukan dengn *Hooks* namun tidak bisa saya lakukan dengan kelas-kelas? {#what-can-i-do-with-hooks-that-i-couldnt-with-classes}

Author: > ### Apa yang dapat saya lakukan dengn *Hooks* namun tidak bisa saya lakukan dengan kelas? 

suggested revision, applied

> Hooks menawarkan sebuah cara baru yang kuat dan ekpresif untuk menggunakan kembali fungsionalitas antara komponen. ["Buat Hooks-mu Sendiri"](/docs/hooks-custom.html) menyajikan sekialas apa saja yang mungkin dilakukan. [Artikel ini](https://medium.com/@dan_abramov/making-sense-of-react-hooks-fdbde8803889), oleh seorang anggota inti tim React, menyelam lebih dalam lagi ke kemampuan-kemampuan baru yang dibuka oleh Hooks.

Author: > Hooks menawarkan sebuah cara baru yang kuat dan ekpresif untuk menggunakan kembali fungsionalitas antara komponen. ["Buat Hooks-mu Sendiri"](/docs/hooks-custom.html) menyajikan sekilas apa saja yang mungkin dilakukan. [Artikel ini](https://medium.com/@dan_abramov/making-sense-of-react-hooks-fdbde8803889), oleh seorang anggota inti tim React, menyelam lebih dalam lagi kemampuan-kemampuan baru yang dibuka oleh Hooks.

suggested revision, applied

Anda tidak bisa menggunakan *Hooks* *di dalam* sebuah komponen kelas, tetapi Anda bisa mencampurkan komponen kelas dan fungsi dengan *Hooks* dalam sebuah tatanan. Tak peduli apakah sebuah komponen adalah kelas atau fungsi yang menggunakan Hooks, itu semua adalah implementasi rinci dari komponen tersebut. Dalam jangka waktu panjang, kami harap *Hooks* menjadi cara utama semua orang untuk menulis komponen React.

### Apakah *Hooks* mencakup seluruh kasus yang digunakan untuk kelas? {#do-hooks-cover-all-use-cases-for-classes}

Tujuan kami untuk *Hooks* yakni mencakup seluruh kasus yang digunakan untuk kelas sesegera mungkin. Tidak ada persamaan (*equivalent*) *Hooks* untuk *lifecyle* yang tidak umum seperti *`getSnapshotBeforeUpdate`* dan *`componentDidCatch`* sementara ini, namun kami berencana untuk menambahkannya segera.

Sekarang adalah masa-masa awal untuk Hooks, dan *library* pihak ketiga bisa jadi tidak kompatibel dengan *Hooks* saat ini.

### Apakah *Hooks* menggantikan *render props* dan *higher-order components*? {#do-hooks-replace-render-props-and-higher-order-components}

> Seringkali, *render props* dan *higher-order components* hanya (me)*render* sebuah turunan. Kami pikir *Hooks* adalah cara yang lebih sederhana untuk menyajikan kasus penggunaan ini. Masih ada tempat untuk kedua pola (contohnya, sebuah *virtual scroller component* bisa saja memiliki sebuah *`renderItem` prop*, atau sebuah *visual container component* bisa saja memiliki struktur *DOM*-nya sendiri). Namun dalam banyak kasus, *Hooks* sudahlah cukup dan akan membantu mengurangi *nesting* dalam tatanan Anda.

Author: >Seringkali, *render props* dan *higher-order components* hanya (me)*render* sebuah turunan. Kami pikir *Hooks* adalah cara yang lebih sederhana untuk menyajikan kasus penggunaan ini. Masih ada tempat untuk kedua pola (contohnya, sebuah *virtual scroller component* bisa saja memiliki sebuah *prop* `renderItem`, atau sebuah *visual container component* bisa saja memiliki struktur *DOM*-nya sendiri). Namun dalam banyak kasus, *Hooks* sudahlah cukup dan akan membantu mengurangi *nesting* dalam tatanan Anda.

suggested revision, applied

### Apa yang *Hooks* peruntukan untuk API populer seperti *Redux `connect()`* dan *React router*? {#what-do-hooks-mean-for-popular-apis-like-redux-connect-and-react-router}

Anda dapat terus menggunakan *API* yang sama seperti yang selama ini dilakukan; *API* tersebut akan tetap bekerja.

*React Redux* sejak versi v7.1.0 [mendukung API Hooks](https://react-redux.js.org/api/hooks) dan memaparkan *Hooks* seperti *`useDispatch`* atau *`useSelector`*.

*Library* seperti *React Router* akan mendukung *Hooks* di masa yang akan datang.

### Apakah *Hooks* dapat berjalan dengan *static typing*?{#do-hooks-work-with-static-typing}

> Hooks didesain dengan *static typing* pada dasarnya. Karena hal-hal tersebut adalah fungsi, *static typing* lebih mudah ditulis dengan benar daripada pola-pola seperti *higher-order components*. Definisi *Flow* dan *TypeScript React* sudah termasuk *support* untuk React *Hooks* di dalamnya.

Author: >*Hooks* didesain dengan *static typing* pada dasarnya. Karena hal-hal tersebut adalah fungsi, *static typing* lebih mudah ditulis dengan benar daripada pola-pola seperti *higher-order components*. Definisi *Flow* dan *TypeScript React* untuk React sudah mendukung untuk React Hooks di dalamnya.

suggested revision, applied

> Penting untuk diingat, *custom Hooks* memberikan Anda kewenangan untuk memaksa *React API* jika Anda ingin menuliskannya dengan lebih ketat dalam beberapa hal. React memberikan Anda *primitives*, tapi Anda bisa menggabungkannya dengan cara-cara lain yang unik selain dari yang kami sediakan untuk Anda.

Reviewer: ```suggestion
Penting untuk diingat, *custom Hooks* memberikan Anda kewenangan untuk memaksa API React jika Anda ingin menuliskannya dengan lebih ketat dalam beberapa hal. React memberikan Anda *primitives*, tapi Anda bisa menggabungkannya dengan cara-cara lain yang unik selain dari yang kami sediakan untuk Anda.
```

Author: >Penting untuk diingat, *custom Hooks* memberikan Anda kewenangan untuk memaksa *API* React jika Anda ingin menuliskannya dengan lebih ketat dalam beberapa hal. React memberikan Anda *primitives*, tapi Anda bisa menggabungkannya dengan cara-cara lain yang unik selain dari yang kami sediakan untuk Anda.

suggested revision, applied

### Bagaimana cara menguji komponen-komponen yang menggunakan Hooks? {#how-to-test-components-that-use-hooks}

Dari sudut pandang React, sebuah komponen yang menggunakan *Hooks* hanyalah sekedar komponen biasa. Jika solusi *testing* Anda tidak bergantung pada fitur-fitur di dalam React, menguji komponen dengan *Hooks* tidak akan berbeda halnya dengan uji komponen yang normalnya Anda lakukan.

>Catatan
>
>[Testing Recipes](/docs/testing-recipes.html) memuat banyak contoh yang bisa Anda salin dan terapkan.

Sebagai contoh, umpamakan kita memiliki komponen penghitung (*counter*) berikut:

```js
function Example() {
  const [count, setCount] = useState(0);
  useEffect(() => {
    document.title = `Anda meng-klik sebanyak ${count} kali`;
  });
  return (
    <div>
      <p>Anda meng-klik sebanyak {count} kali</p>
      <button onClick={() => setCount(count + 1)}>
        Klik sayaS
      </button>
    </div>
  );
}
```

> Kami akan mengujinya dengan React *DOM*. Pastikan bahwa perilakunya dengan yang terjadi di peramban, kami akan mengemas proses *rendering* kode dan memperbaruinya ke dalam [`ReactTestUtils.act()`](/docs/test-utils.html#act) beberapa pemanggilan:

Reviewer: ```suggestion
Kami akan mengujinya dengan React DOM. Pastikan bahwa perilakunya dengan yang terjadi di peramban, kami akan mengemas proses *rendering* kode dan memperbaruinya ke dalam [`ReactTestUtils.act()`](/docs/test-utils.html#act) beberapa pemanggilan:
```

Abbreviations in foreign languages need not be italicised.
Reference: https://twitter.com/ivanlanin/status/898520073224167424

Author: >Kami akan mengujinya dengan React DOM. Pastikan bahwa perilakunya dengan yang terjadi di peramban, kami akan mengemas proses *rendering* kode dan memperbaruinya ke dalam [`ReactTestUtils.act()`](/docs/test-utils.html#act) beberapa pemanggilan:

suggested revision, applied

  * [Bagaimana cara menguji komponen-komponen yang menggunakan Hooks?](#how-to-test-components-that-use-hooks)
  * [Apa yang sebenarnya aturan-aturan *lint* terapkan?](#what-exactly-do-the-lint-rules-enforce)
* **[Dari Kelas ke Hooks](#from-classes-to-hooks)**
  * [Bagaimana cara *lifecycle method* dapat sesuai dengan Hooks?](#how-do-lifecycle-methods-correspond-to-hooks)
  * [Bagaimana cara saya dapat memperoleh data dengan Hooks?](#how-can-i-do-data-fetching-with-hooks)
  * [Apakah ada hal seperti *instance variable*?](#is-there-something-like-instance-variables)
  * [Apa saya harus menggunakan satu atau beberapa *state variable*?](#should-i-use-one-or-many-state-variables)
  * [Bisakah saya menjalakan sebuah efek hanya pada pembaruan?](#can-i-run-an-effect-only-on-updates)
  * [Bagimana cara untuk mendapatkan kembali *prop* atau *state* sebelumnya?](#how-to-get-the-previous-props-or-state)
  * [Mengapa saya melihat *prop* atau *state* yang basi dalam fungsi yang saya buat?](#why-am-i-seeing-stale-props-or-state-inside-my-function)
  * [Bagaimana cara mengimplementasikan *getDerivedStateFromProps*?](#how-do-i-implement-getderivedstatefromprops)

>   * [Apakah ada suatu hal seperti *forceUpdate*?](#is-there-something-like-forceupdate)

Reviewer: ```suggestion
  * [Apakah ada fungsionalitas seperti *forceUpdate*?](#is-there-something-like-forceupdate)
```

  * [Bisakah saya membuat sebuah *ref* pada sebuah komponen fungsi?](#can-i-make-a-ref-to-a-function-component)

>   * [Bagaimana cara mengukur sebuah node *DOM*](#how-can-i-measure-a-dom-node)

Reviewer: ```suggestion
  * [Bagaimana cara mengukur sebuah simpul DOM?](#how-can-i-measure-a-dom-node)
```

Hooks menawarkan sebuah cara baru yang kuat dan ekpresif untuk menggunakan kembali fungsionalitas antara komponen. ["Buat Hooks-mu Sendiri"](/docs/hooks-custom.html) menyajikan sekilas apa saja yang mungkin dilakukan. [Artikel ini](https://medium.com/@dan_abramov/making-sense-of-react-hooks-fdbde8803889), oleh seorang anggota inti tim React, menyelam lebih dalam lagi kemampuan-kemampuan baru yang dibuka oleh Hooks.

### Seberapa banyak pengetahuan saya tentang React yang tetap relevan? {#how-much-of-my-react-knowledge-stays-relevant}

Hooks adalah cara yang lebih langsung untuk menggunakan fitur-fitur React yang sudah Anda ketahui -- seperti *state, lifecycle, context,* dan *ref*. Hal-hal tersebut tidak secara mendasar merubah bagaimana React bekerja, dan pengetahuan Anda tentang komponen, *props*, dan *top-down data flow* tetaplah relevan.

Hooks memiliki sebuah alur kurva pembelajarannya sendiri. Jika terdapat sesuatu yang hilang dalam dokumentasi ini, [ajukan isu](https://github.com/reactjs/reactjs.org/issues/new) dan kami akan coba membantu.

### Haruskah saya menggunakan Hooks, kelas atau gabungan keduanya?{#should-i-use-hooks-classes-or-a-mix-of-both}

> Ketika Anda siap, kami menganjurkan Anda untuk mulai mencoba *Hooks* dalam komponen-komponen baru yang Anda tulis. Pastikan setiap orang dalam tim Anda siap dan tidak asing dengan dokumentasi ini. Kami tidak merekomendasikan menulis ulang kelas-kelas Anda yang sudah ada ke *Hooks* kecuali Anda sudah berencana menulis ulang (seperti, memperbaiki *bug*).

Reviewer: ```suggestion
Ketika Anda siap, kami menganjurkan Anda untuk mulai mencoba *Hooks* dalam komponen-komponen baru yang Anda tulis. Pastikan setiap orang dalam tim Anda siap dan tidak asing dengan dokumentasi ini. Kami tidak merekomendasikan menulis ulang kelas-kelas Anda yang sudah ada ke *Hooks* kecuali Anda sudah berencana menulis ulang (seperti memperbaiki *bug*).
```

No need the comma here.

Author: >Ketika Anda siap, kami menganjurkan Anda untuk mulai mencoba *Hooks* dalam komponen-komponen baru yang Anda tulis. Pastikan setiap orang dalam tim Anda siap dan tidak asing dengan dokumentasi ini. Kami tidak merekomendasikan menulis ulang kelas-kelas Anda yang sudah ada ke *Hooks* kecuali Anda sudah berencana menulis ulang (seperti memperbaiki *bug*).

Suggested change's applied

Seringkali, *render props* dan *higher-order components* hanya (me)*render* sebuah turunan. Kami pikir *Hooks* adalah cara yang lebih sederhana untuk menyajikan kasus penggunaan ini. Masih ada tempat untuk kedua pola (contohnya, sebuah *virtual scroller component* bisa saja memiliki sebuah *prop* `renderItem`, atau sebuah *visual container component* bisa saja memiliki struktur *DOM*-nya sendiri). Namun dalam banyak kasus, *Hooks* sudahlah cukup dan akan membantu mengurangi *nesting* dalam tatanan Anda.

### Apa kegunaan *Hooks* untuk API populer seperti Redux `connect()` dan React router? {#what-do-hooks-mean-for-popular-apis-like-redux-connect-and-react-router}

Anda dapat terus menggunakan API yang sama seperti yang selama ini dilakukan; API tersebut akan tetap bekerja.

React Redux sejak versi v7.1.0 [mendukung API Hooks](https://react-redux.js.org/api/hooks) dan memaparkan *Hooks* seperti `useDispatch` atau `useSelector`.

*Library* seperti React Router akan mendukung *Hooks* di masa yang akan datang.

### Apakah *Hooks* dapat bekerja dengan *static typing*?{#do-hooks-work-with-static-typing}

> *Hooks* didesain dengan *static typing* pada dasarnya. Karena *Hooks* adalah sekumpulan fungsi, *static typing* lebih mudah ditulis dengan benar daripada pola-pola seperti *higher-order components*. Definisi Flow dan TypeScript React untuk React sudah mendukung React Hooks.

Reviewer: ```suggestion
*Hooks* didesain dengan *static typing* pada dasarnya. Karena *Hooks* adalah sekumpulan fungsi, *static typing* lebih mudah ditulis dengan benar daripada pola-pola seperti *higher-order components*. Definisi Flow dan TypeScript untuk React sudah mendukung React Hooks.
```

Double mention of React in here, perhaps?

Author: >*Hooks* didesain dengan *static typing* pada dasarnya. Karena *Hooks* adalah sekumpulan fungsi, *static typing* lebih mudah ditulis dengan benar daripada pola-pola seperti *higher-order components*. Definisi Flow dan TypeScript untuk React sudah mendukung React Hooks.

Suggested change's applied

### Bagaimana cara menguji komponen-komponen yang menggunakan Hooks? {#how-to-test-components-that-use-hooks}

Dari sudut pandang React, sebuah komponen yang menggunakan *Hooks* hanyalah sekedar komponen biasa. Jika solusi *testing* Anda tidak bergantung pada fitur-fitur di dalam React, menguji komponen dengan *Hooks* tidak akan berbeda halnya dengan uji komponen yang normalnya Anda lakukan.

>Catatan
>
>[Testing Recipes](/docs/testing-recipes.html) memuat banyak contoh yang bisa Anda salin dan terapkan.

Sebagai contoh, umpamakan kita memiliki komponen penghitung (*counter*) berikut:

```js
function Example() {
  const [count, setCount] = useState(0);
  useEffect(() => {
    document.title = `Anda meng-klik sebanyak ${count} kali`;
  });
  return (
    <div>
      <p>Anda meng-klik sebanyak {count} kali</p>
      <button onClick={() => setCount(count + 1)}>
        Klik saya
      </button>
    </div>
  );
}
```

> Kami akan mengujinya dengan React DOM. Pastikan bahwa perilakunya dengan yang terjadi di peramban, kami akan mengemas proses *rendering* kode dan memperbaruinya ke dalam [`ReactTestUtils.act()`](/docs/test-utils.html#act) beberapa pemanggilan:

Reviewer: ```suggestion

Kami akan mengujinya dengan React DOM. Pastikan bahwa perilakunya dengan yang terjadi di peramban, kami akan mengemas proses *rendering* kode dan memperbaruinya ke dalam [`ReactTestUtils.act()`](/docs/test-utils.html#act) beberapa pemanggilan:
```

Missing paragraph break here. This will cause merge conflicts when we sync new updates from the upstream reactjs.org docs.

Author: Bring back missing paragraph break

<!--
  jika Anda perlu me-rgenerate ini, snippet dalam konsol devtools ini bisa membantu:

  $$('.anchor').map(a =>
    `${' '.repeat(2 * +a.parentNode.nodeName.slice(1))}` +
    `[${a.parentNode.textContent}](${a.getAttribute('href')})`
  ).join('\n')
-->

* **[Strategi Adopsi](#adoption-strategy)**
  * [Versi React mana yang termasuk *Hooks* di dalamnya?](#which-versions-of-react-include-hooks)
  * [Apakah perlu menulis ulang semua komponen kelas saya?](#do-i-need-to-rewrite-all-my-class-components)
  * [Apa yang dapat saya lakukan dengan *Hooks* namun tidak bisa saya lakukan dengan kelas?](#what-can-i-do-with-hooks-that-i-couldnt-with-classes)
  * [Seberapa banyak pengetahuan saya tentang React yang tetap relevan?](#how-much-of-my-react-knowledge-stays-relevant)
  * [Haruskah saya menggunakan Hooks, kelas atau gabungan keduanya?](#should-i-use-hooks-classes-or-a-mix-of-both)
  * [Apakah *Hooks* mencakup seluruh kasus yang digunakan untuk kelas?](#do-hooks-cover-all-use-cases-for-classes)
  * [Apakah *Hooks* menggantikan *render props* dan *higher-order components*?](#do-hooks-replace-render-props-and-higher-order-components)

>   * [Apa yang *Hooks* peruntukan untuk API populer seperti *Redux connect()* dan *React router*?](#what-do-hooks-mean-for-popular-apis-like-redux-connect-and-react-router)

Reviewer: ```suggestion
  * [Apakah kegunaan *Hooks* untuk API populer seperti Redux connect() dan React router?](#what-do-hooks-mean-for-popular-apis-like-redux-connect-and-react-router)
```

- Library names/names of things do not need to be italicised.
- This translation slightly makes more sense.

Author: >* [Apa kegunaan *Hooks* untuk API populer seperti Redux connect() dan React router?](#what-do-hooks-mean-for-popular-apis-like-redux-connect-and-react-router)

Suggested changes are applied

>   * [Apakah *Hooks* dapat berjalan dengan *static typing*?](#do-hooks-work-with-static-typing)

Reviewer: ```suggestion
  * [Apakah *Hooks* dapat bekerja dengan *static typing*?](#do-hooks-work-with-static-typing)
```

Author: >   * [Apakah *Hooks* dapat bekerja dengan *static typing*?](#do-hooks-work-with-static-typing)

Suggested change's applied

  * [Bagaimana cara menguji komponen-komponen yang menggunakan Hooks?](#how-to-test-components-that-use-hooks)
  * [Apa yang sebenarnya aturan-aturan *lint* terapkan?](#what-exactly-do-the-lint-rules-enforce)
* **[Dari Kelas ke Hooks](#from-classes-to-hooks)**
  * [Bagaimana cara *lifecycle method* dapat sesuai dengan Hooks?](#how-do-lifecycle-methods-correspond-to-hooks)
  * [Bagaimana cara saya dapat memperoleh data dengan Hooks?](#how-can-i-do-data-fetching-with-hooks)
  * [Apakah ada hal seperti *instance variable*?](#is-there-something-like-instance-variables)
  * [Apa saya harus menggunakan satu atau beberapa *state variable*?](#should-i-use-one-or-many-state-variables)
  * [Bisakah saya menjalakan sebuah efek hanya pada pembaruan?](#can-i-run-an-effect-only-on-updates)
  * [Bagimana cara untuk mendapatkan kembali *prop* atau *state* sebelumnya?](#how-to-get-the-previous-props-or-state)

>   * [Mengapa saya melihat *stale prop* atau *state* dalam fungsi yang saya buat?](#why-am-i-seeing-stale-props-or-state-inside-my-function)

Reviewer: ```suggestion
  * [Mengapa saya melihat *prop* atau *state* yang basi dalam fungsi yang saya buat?](#why-am-i-seeing-stale-props-or-state-inside-my-function)
```

"stale" might be translatable to "basi" so let's go with that.

Author: >   * [Mengapa saya melihat *prop* atau *state* yang basi dalam fungsi yang saya buat?](#why-am-i-seeing-stale-props-or-state-inside-my-function)

Suggested change's applied

Author: > "stale" might be translatable to "basi" so let's go with that.

Understood. Thank you

### Apakah *Hooks* mencakup seluruh kasus yang digunakan untuk kelas? {#do-hooks-cover-all-use-cases-for-classes}

Tujuan kami untuk *Hooks* yakni mencakup seluruh kasus yang digunakan untuk kelas sesegera mungkin. Tidak ada persamaan (*equivalent*) *Hooks* untuk *lifecyle* yang tidak umum seperti `getSnapshotBeforeUpdate` dan `componentDidCatch` sementara ini, namun kami berencana untuk menambahkannya segera.

Sekarang adalah masa-masa awal untuk Hooks, dan *library* pihak ketiga bisa jadi tidak kompatibel dengan *Hooks* saat ini.

### Apakah *Hooks* menggantikan *render props* dan *higher-order components*? {#do-hooks-replace-render-props-and-higher-order-components}

Seringkali, *render props* dan *higher-order components* hanya (me)*render* sebuah turunan. Kami pikir *Hooks* adalah cara yang lebih sederhana untuk menyajikan kasus penggunaan ini. Masih ada tempat untuk kedua pola (contohnya, sebuah *virtual scroller component* bisa saja memiliki sebuah *prop* `renderItem`, atau sebuah *visual container component* bisa saja memiliki struktur *DOM*-nya sendiri). Namun dalam banyak kasus, *Hooks* sudahlah cukup dan akan membantu mengurangi *nesting* dalam tatanan Anda.

> ### Apa yang *Hooks* peruntukan untuk API populer seperti *Redux `connect()` dan *React router*? {#what-do-hooks-mean-for-popular-apis-like-redux-connect-and-react-router}

Reviewer: ```suggestion
### Apakah kegunaan *Hooks* untuk API populer seperti Redux `connect()` dan React router? {#what-do-hooks-mean-for-popular-apis-like-redux-connect-and-react-router}
```

Idem.

Author: > ### Apa kegunaan *Hooks* untuk API populer seperti Redux `connect()` dan React router? {#what-do-hooks-mean-for-popular-apis-like-redux-connect-and-react-router}

Suggested changes are applied

> Anda dapat terus menggunakan *API* yang sama seperti yang selama ini dilakukan; *API* tersebut akan tetap bekerja.

Reviewer: ```suggestion
Anda dapat terus menggunakan API yang sama seperti yang selama ini dilakukan; API tersebut akan tetap bekerja.
```

Please remove all italicisation of API. Abbreviations for foreign names don't need to be translated.

https://twitter.com/ivanlanin/status/898520073224167424

Author: > Anda dapat terus menggunakan API yang sama seperti yang selama ini dilakukan; API tersebut akan tetap bekerja.

Suggested change's applied

Author: > Please remove all italicisation of API. Abbreviations for foreign names don't need to be translated.
> 
> https://twitter.com/ivanlanin/status/898520073224167424

All italicization of `API` are changed.

React Redux sejak versi v7.1.0 [mendukung API Hooks](https://react-redux.js.org/api/hooks) dan memaparkan *Hooks* seperti `useDispatch` atau `useSelector`.

*Library* seperti React Router akan mendukung *Hooks* di masa yang akan datang.

> ### Apakah *Hooks* dapat berjalan dengan *static typing*?{#do-hooks-work-with-static-typing}

Reviewer: ```suggestion
### Apakah *Hooks* dapat bekerja dengan *static typing*?{#do-hooks-work-with-static-typing}
```

Author: > ### Apakah *Hooks* dapat bekerja dengan *static typing*?{#do-hooks-work-with-static-typing}

Suggested change's applied

> *Hooks* didesain dengan *static typing* pada dasarnya. Karena hal-hal tersebut adalah fungsi, *static typing* lebih mudah ditulis dengan benar daripada pola-pola seperti *higher-order components*. Definisi *Flow* dan *TypeScript React* untuk React sudah mendukung untuk React Hooks di dalamnya.

Reviewer: ```suggestion
*Hooks* didesain dengan *static typing* pada dasarnya. Karena *Hooks* adalah sekumpulan fungsi, *static typing* lebih mudah ditulis dengan benar daripada pola-pola seperti *higher-order components*. Definisi Flow dan TypeScript untuk React sudah mendukung React Hooks.
```

- Ambiguous phrasing eliminated.
- Names of other languages (e.g. Flow and TypeScript) don't need to be italicised

Author: >*Hooks* didesain dengan *static typing* pada dasarnya. Karena *Hooks* adalah sekumpulan fungsi, *static typing* lebih mudah ditulis dengan benar daripada pola-pola seperti *higher-order components*. Definisi Flow dan TypeScript React untuk React sudah mendukung React Hooks.

Suggested changes are applied.

Untuk informasi lebih lanjut, cek [Testing Recipes](/docs/testing-recipes.html).

### Apa yang sebenarnya [aturan lint](https://www.npmjs.com/package/eslint-plugin-react-hooks) terapkan? {#what-exactly-do-the-lint-rules-enforce}

Kami menyediakan sebuah [*plugin* ESLint](https://www.npmjs.com/package/eslint-plugin-react-hooks) yang menerapkan [aturan-aturan Hooks](/docs/hooks-rules.html) demi menghindari *bug*. *Plugin* tersebut mengasumsikan bahwa fungsi apapun yang diawali dengan "`use`" dan sebuah huruf kapital tepat setelahnya, adalah sebuah Hook. Kami menyadari pendekatan penemuan (heuristik) ini tidaklah sempurna dan bisa jadi masih ada beberapa *false positive*, akan tetapi tanpa sebuah perjanjian (konvensi) seluruh lingkup ekosistem tidak akan mungkin *Hooks* akan berjalan dengan baik -- dan nama-nama yang lebih panjang akan mengecilkan minat orang-orang baik untuk mengadopsi *Hooks* maupun mengikuti konvensinya.

Khususnya, aturan tersebut menekankan bahwa:

> * Panggilan untuk *Hooks* bisa terdapat di dalam sebuah fungsi *`PascalCase`* (diasumsikan sebagai sebuah komponen) atau fungsi *`useSomething`* lainnya (diasumsikan sebagai sebuah *custom Hook*).

Reviewer: ```suggestion
* Panggilan untuk *Hooks* bisa terdapat di dalam sebuah fungsi *`PascalCase`* (diasumsikan sebagai sebuah komponen) atau fungsi `useSomething` lainnya (diasumsikan sebagai sebuah *custom Hook*).
```

Method names inside code blocks don't need to be italicised.

* *`getDerivedStateFromProps`*: Jadwalkan sebuah pembaruan [ketika proses *render*](#how-do-i-implement-getderivedstatefromprops).

* *`shouldComponentUpdate`*: Lihat `React.memo` [di bawah ini](#how-do-i-implement-shouldcomponentupdate).

* *`render`*: Ini adalah komponen fungsi *body* itu sendiri.

* *`componentDidMount`, `componentDidUpdate`, `componentWillUnmount`*: [*`useEffect` Hook*](/docs/hooks-reference.html#useeffect) dapat mengekspresikan semua kombinasi ini (termasuk kasus-kasus yang [kurang](#can-i-skip-an-effect-on-updates) [awam](#can-i-run-an-effect-only-on-updates)).

* *`componentDidCatch`* dan *`getDerivedStateFromError`*: Tidak ada persamaan Hook untuk *method* ini sementara ini, tetapi akan segera ditambahkan.

### Bagaimana cara saya dapat memperoleh data dengan Hooks?{#how-can-i-do-data-fetching-with-hooks}

Berikut ini sebuah [demo kecil](https://codesandbox.io/s/jvvkoo8pq3) untuk permulaan Anda. Untuk mempelajari lebih lanjut, lihat [attikel ini](https://www.robinwieruch.de/react-hooks-fetch-data/) tentang memperoleh (*fetching*) data dengan menggunakan Hooks.

### Apakah ada hal seperti *instance variable*? {#is-there-something-like-instance-variables}

> Ya! [`useRef()`](/docs/hooks-reference.html#useref) Hook tidak hanya untuk *DOM ref*. Objek *"ref"* adalah sebuah *container* umum yang mana memiliki properti *`current`* yang berubah-ubah serta dapat memegang nilai apapun, serupa dengan *instance property* pada sebuah kelas.

Reviewer: ```suggestion
Ya! [`useRef()`](/docs/hooks-reference.html#useref) Hook tidak hanya untuk *ref* DOM. Objek *"ref"* adalah sebuah *container* umum yang mana memiliki properti `current` yang berubah-ubah serta dapat memegang nilai apapun, serupa dengan *instance property* pada sebuah kelas.
```

Author: >Ya! [`useRef()`](/docs/hooks-reference.html#useref) Hook tidak hanya untuk *ref* DOM. Objek *"ref"* adalah sebuah *container* umum yang mana memiliki properti `current` yang berubah-ubah serta dapat memegang nilai apapun, serupa dengan *instance property* pada sebuah kelas.

suggested revision, applied

  // ...
```

> Secara konsep, Anda bisa pikirkan *ref* itu serupa dengan *instance variables* dalam sebuah kelas. Kecuali, Anda menerapkan [lazy initialization](#how-to-create-expensive-objects-lazily), hindari pengaturan ref selama proses *render* -- hal ini akan mengakibatkan *behavior* yang tidak terduga. Sebagai gantinya, biasanya Anda ingin untuk memodifikasi *ref* dalam *event handler* dan efek.

Reviewer: ```suggestion
Secara konsep, Anda bisa pikirkan *ref* itu serupa dengan variabel *instance* dalam sebuah kelas. Kecuali, Anda menerapkan [lazy initialization](#how-to-create-expensive-objects-lazily), hindari pengaturan ref selama proses *render* -- hal ini akan mengakibatkan *behavior* yang tidak terduga. Sebagai gantinya, biasanya Anda ingin untuk memodifikasi *ref* dalam *event handler* dan efek.
```

Author: >Secara konsep, Anda bisa pikirkan *ref* itu serupa dengan variabel *instance* dalam sebuah kelas. Kecuali, Anda menerapkan [lazy initialization](#how-to-create-expensive-objects-lazily), hindari pengaturan ref selama proses *render* -- hal ini akan mengakibatkan *behavior* yang tidak terduga. Sebagai gantinya, biasanya Anda ingin untuk memodifikasi *ref* dalam *event handler* dan efek.

Suggested change's applied

}
```

Perhatikan bagaimana kita dapat memindahkan panggilan `useState` untuk *state variable* `position` serta efek yang terhubung ke dalam sebuah *custom Hook* tanpa merubah kodenya. Jika semua *state* ada dalam sebuah objek, mengekstrak *state* tersebut akan jadi lebih sulit.

Baik memasang semua *state* ke dalam sebuah panggilan `useState`, serta memiliki sebuah panggilan `useState` pada setiap area, keduanya sama-sama berfungsi. Komponen-komponen cenderung paling mudah dibaca ketika Anda menemukan keseimbangan antara (penerapan) kedua cara tersebut, dan kelompok yang terhubung dengan *state* ke dalam beberapa *state variable* independen. Jika logika *state* tersebut menjadi kompleks, kami rekomendasikan [mengelolanya dengan sebuah *reducer*](/docs/hooks-reference.html#usereducer) atau sebuah *custom Hook*.

### Bisakah saya menjalakan sebuah efek hanya pada pembaruan? {#can-i-run-an-effect-only-on-updates}

> Ini adalah kasus penggunaan yang langka. Jika butuh, Anda bisa [gunakan sebuah *ref* yang bisa diubah-ubah](#is-there-something-like-instance-variables) menjadi secara manual menyetorkan nilai *boolean* sesuai dengan apakah Anda berada pada *render* pertama atau selanjutnya, kemudia memeriksa penandanya (*flag*) pada efek yang Anda buat. (Jika sering melakukan hal ini, ada baiknya Anda membuat sebuah *custom Hook*.)

Reviewer: ```suggestion
Ini adalah kasus penggunaan yang langka. Jika butuh, Anda dapat [menggunakan sebuah *ref* yang *mutable*](#is-there-something-like-instance-variables) untuk menyimpan secara manual nilai *boolean* sesuai dengan apakah Anda berada pada *render* pertama atau selanjutnya, kemudian memeriksa penandanya (*flag*) pada efek yang Anda buat. (Jika sering melakukan hal ini, ada baiknya Anda membuat sebuah *custom Hook*.)
```

This translation makes more sense, IMO.

Author: >Ini adalah kasus penggunaan yang langka. Jika butuh, Anda dapat [menggunakan sebuah *ref* yang bisa *mutable*](#is-there-something-like-instance-variables)untuk menyimpan secara manual nilai *boolean* sesuai dengan apakah Anda berada pada *render* pertama atau selanjutnya, kemudian memeriksa penandanya (*flag*) pada efek yang Anda buat. (Jika sering melakukan hal ini, ada baiknya Anda membuat sebuah *custom Hook*.)

Suggested changes are applied

### Bagaimana cara untuk mendapatkan kembali *prop* atau *state* sebelumnya? {#how-to-get-the-previous-props-or-state}

> Kini, Anda bisa melakukannya secara manual [dengan *ref*](#is-there-something-like-instance-variables):

Reviewer: ```suggestion
Saat ini, Anda dapat melakukannya secara manual [dengan *ref*](#is-there-something-like-instance-variables):
```

Author: >Saat ini, Anda dapat melakukannya secara manual [dengan *ref*](#is-there-something-like-instance-variables):

Suggested changes are applied

}
```

> Ini bisa jadi agak berbelit tetapi Anda bisa mengekstraknya ke dalam sebuah *custom Hook*:

Reviewer: ```suggestion
Cara ini mungkin agak rumit tetapi Anda bisa mengekstraknya ke dalam sebuah *custom Hook*:
```

Author: >Cara ini mungkin agak rumit tetapi Anda bisa mengekstraknya ke dalam sebuah *custom Hook*:

Suggested changes are applied

}
```

> Perhatikan bagaimana hal ini dapat bekerja pada *props, state,* atau nilai apapun yang diperhitungkan.

Reviewer: ```suggestion
Perhatikan bagaimana *Hook* ini dapat bekerja pada *props, state,* atau nilai apapun yang diperhitungkan.
```

Author: >Perhatikan bagaimana *Hook* ini dapat bekerja pada *props, state,* atau nilai apapun yang diperhitungkan.

Suggested change's applied

  // ...
```

> Di masa yang akan datang, React mungkin akan menyediakan sebuah `usePrevious` Hook* yang unik mengingat hal tersebut adalah kasus yang cukup sering digunakan.

Reviewer: ```suggestion
Di masa yang akan datang, React mungkin akan menyediakan sebuah *Hook* `usePrevious` mengingat hal tersebut adalah kasus yang cukup sering digunakan.
```

Author: >Di masa yang akan datang, React mungkin akan menyediakan sebuah *Hook* `usePrevious` mengingat hal tersebut adalah kasus yang cukup sering digunakan.

Suggested changes are applied

Lihat juga [pola yang direkomendasikan untuk *derived state*](#how-do-i-implement-getderivedstatefromprops).

> ### Mengapa saya melihat *stale prop* atau *state* dalam fungsi yang saya buat?  {#why-am-i-seeing-stale-props-or-state-inside-my-function}

Reviewer: ```suggestion
### Mengapa saya melihat *prop* atau *state* yang basi dalam fungsi yang saya buat?  {#why-am-i-seeing-stale-props-or-state-inside-my-function}
```

Author: > ### Mengapa saya melihat *prop* atau *state* yang basi dalam fungsi yang saya buat?  {#why-am-i-seeing-stale-props-or-state-inside-my-function}

Suggested change's applied

Jika ingin membaca *state* *terkini* dari beberapa *asynchronous callback*, baiknya Anda menyimpannya dalam [sebuah *ref*](/docs/hooks-faq.html#is-there-something-like-instance-variables), mengubahnya, dan membaca dari situ.

Pada akhirnya, alasan lain Anda melihat *prop* atau *state* yang basi yakni jika Anda menggunakan optimasi "dependency array" tapi tidak secara benar dalam menspesifikasi semua *dependency*-nya. Sebagai contohnya, jika sebuah efek menspesifikasikan `[]` sebagai argumen kedua tetapi membaca `someProp` di dalamnya, efek tersebut akan tetap "melihat" nilai awal dari `someProp`. Solusinya yakni antara menghilangkan *dependency array*, atau memperbaikinya. Ini adalah cara [bagaimana Anda bisa memperlakukan fungsi](#is-it-safe-to-omit-functions-from-the-list-of-dependencies), dan ini adalah [strategi umum lainnya](#what-can-i-do-if-my-effect-dependencies-change-too-often) untuk menjalankan efek lebih jarang tanpa keliru melewatkan *dependency*.

>Catatan
>

> >Kami menyediakan sebuah aturan *ESLint* [`exhaustive-deps`](https://github.com/facebook/react/issues/14920) sebagai bagian dari *package*[`eslint-plugin-react-hooks`](https://www.npmjs.com/package/eslint-plugin-react-hooks#installation). Itu akan memberi peringatan ketika *dependency* keliru dispesifikasikan dan akan memberikan saran perbaikan.

Reviewer: ```suggestion
>Kami menyediakan sebuah aturan *ESLint* [`exhaustive-deps`](https://github.com/facebook/react/issues/14920) sebagai bagian dari *package*[`eslint-plugin-react-hooks`](https://www.npmjs.com/package/eslint-plugin-react-hooks#installation). Aturan ini akan memberi peringatan ketika *dependency* keliru dispesifikasikan dan akan memberikan saran perbaikan.
```

Ambiguous "itu".

Author: > >Kami menyediakan sebuah aturan *ESLint* [`exhaustive-deps`](https://github.com/facebook/react/issues/14920) sebagai bagian dari *package*[`eslint-plugin-react-hooks`](https://www.npmjs.com/package/eslint-plugin-react-hooks#installation). Aturan ini akan memberi peringatan ketika *dependency* keliru dispesifikasikan dan akan memberikan saran perbaikan.

Suggested correction's applied

}
```

Hal ini bisa jadi terlihat aneh pada awalnya, tetapi sebuah pembaruan selama proses *render* adalah hal yang persis secara konsep sejak `getDerivedStateFromProps` dibuat.

> ### Apakah ada suatu hal seperti *forceUpdate*? {#is-there-something-like-forceupdate}

Reviewer: ```suggestion
### Apakah ada fungsionalitas seperti *forceUpdate*? {#is-there-something-like-forceupdate}
```

Author: >### Apakah ada fungsionalitas seperti *forceUpdate*? {#is-there-something-like-forceupdate}

Suggested change's applied.
Table of contents also updated

  }
```

Coba hindari pola berikut ini sebisa mungkin.

### Bisakah saya membuat *ref* pada sebuah komponen fungsi? {#can-i-make-a-ref-to-a-function-component}

Saat Anda seharusnya tidak sering memerlukan hal ini, Anda bisa saja mengekspos beberapa metode perintah pada sebuah *parent component* dengan menggunakan [`useImperativeHandle`](/docs/hooks-reference.html#useimperativehandle) *Hook*.

> ### Bagaimana cara mengukur *DOM node*? {#how-can-i-measure-a-dom-node}

Reviewer: ```suggestion
### Bagaimana cara mengukur sebuah simpul DOM? {#how-can-i-measure-a-dom-node}
```

Author: >### Bagaimana cara mengukur *node* DOM? {#how-can-i-measure-a-dom-node}

I keep the 'node' still as 'node', but remove the italicization of DOM.

> Untuk mengukur posisi atau ukuran dari sebuah *node* DOM, Anda bisa gunakan [*callback ref*](/docs/refs-and-the-dom.html#callback-refs). React akan memanggil *callback* itu kapanpun *ref* terikat pada sebuah *node* yang berbeda. Berikut ini [demo kecil](https://codesandbox.io/s/l7m0v5x4v9):

Reviewer: ```suggestion
Untuk mengukur posisi atau ukuran dari sebuah simpul DOM, Anda bisa gunakan [*callback ref*](/docs/refs-and-the-dom.html#callback-refs). React akan memanggil *callback* itu kapanpun *ref* terikat pada sebuah *node* yang berbeda. Berikut ini [demo kecil](https://codesandbox.io/s/l7m0v5x4v9):
```

Author: >* [Bagaimana cara mengukur sebuah node *DOM*](#how-can-i-measure-a-dom-node)

  }

  doSomething();
}, []); // ✅ OK dalam contoh ini karena kita tidak menggunakan nilai apapun dari cakupan komponen tersebut
```

Tergantung pada kasus penggunaan Anda, terdapat beberapa pilihan lainnya sebagaimana yang dideskripsikan di bawah ini.

>Catatan
>

> >Kami menyediakan aturan *ESLint* [`exhaustive-deps`](https://github.com/facebook/react/issues/14920) sebagai bagian dari *package* [`eslint-plugin-react-hooks`](https://www.npmjs.com/package/eslint-plugin-react-hooks#installation). Itu akan membantu Anda menemukan komponen yang tidak konsisten menangani pembaruan.

Reviewer: ```suggestion
>Kami menyediakan aturan *ESLint* [`exhaustive-deps`](https://github.com/facebook/react/issues/14920) sebagai bagian dari *package* [`eslint-plugin-react-hooks`](https://www.npmjs.com/package/eslint-plugin-react-hooks#installation). Aturan ini akan membantu Anda menemukan komponen yang tidak konsisten menangani pembaruan.
```

Ambiguous "itu".

Author: >>Kami menyediakan aturan *ESLint* [`exhaustive-deps`](https://github.com/facebook/react/issues/14920) sebagai bagian dari *package* [`eslint-plugin-react-hooks`](https://www.npmjs.com/package/eslint-plugin-react-hooks#installation). Aturan ini akan membantu Anda menemukan komponen yang tidak konsisten menangani pembaruan.

Suggested change's applied

### Bagaimana cara mengimplementasikan `shouldComponentUpdate`? {#how-do-i-implement-shouldcomponentupdate}

Anda dapat mengemas sebuah komponen fungsi dengan `React.memo` untuk membandingkan secara tak mendalam *prop*-nya:

```js
const Button = React.memo((props) => {
  // komponen Anda
});
```

Bukanlah sebuah *Hook* karena hal tersebut tidak melakukan penyusunan seperti yang *Hooks* lakukan. `React.memo` adalah persamaan untuk `PureComponent`, tapi hanya untuk membandingkan *prop*. (Anda juga bisa menambahkan argumen kedua untuk menspesifikasikan sebuah fungsi perbandingan yang menerima *prop* lama dan *prop* baru. Jika *return* dari fungsi tersebut adalah *true*, maka pembaruan akan dilewat.)

`React.memo` tidak membandingkan *state* karena tidak ada satu pun objek *state* untuk dibandingkan. Tapi Anda dapat membuat *children* (turunan) murni pula, atau bahkan [mengoptimasi masing-masing *children* dengan menggunakan `useMemo`](/docs/hooks-faq.html#how-to-memoize-calculations).

### Bagaimana cara *memoize* perhitungan? {#how-to-memoize-calculations}

> *Hook* [`useMemo`](/docs/hooks-reference.html#usememo) mengizinkan Anda untuk menyimpan (*cache*) perhitungan-perhitungan di antara berbagai *render* dengan cara *"mengingat"* komputasi sebelumnya:

Reviewer: ```suggestion
*Hook* [`useMemo`](/docs/hooks-reference.html#usememo) memungkinkan Anda untuk menyimpan (*cache*) perhitungan-perhitungan di antara berbagai *render* dengan cara *"mengingat"* komputasi sebelumnya:
```

Author: >*Hook* [`useMemo`](/docs/hooks-reference.html#usememo) memungkinkan Anda untuk menyimpan (*cache*) perhitungan-perhitungan di antara berbagai *render* dengan cara *"mengingat"* komputasi sebelumnya:

Suggested change's applied

Secara tradisional, performa menyangkut seputar fugsi-fungsi *inline* React sudah dihubungkan dengan bagaimana cara untuk mengoper *callback* baru setiap *render* yang memotong optimasi `shouldComponentUpdate` dalam komponen turunan. *Hooks* mendekati masalah dari tiga sisi.

* *Hook* [`useCallback`](/docs/hooks-reference.html#usecallback) mengizinkan Anda untuk tetap menggunakan rujukan *callback* yang sama di antara *render* ulang sehingga `shouldComponentUpdate` akan terus bekerja:

    ```js{2}
    // Tidak akan berubah kecuali `a` atau `b` berubah
    const memoizedCallback = useCallback(() => {
      doSomething(a, b);
    }, [a, b]);
    ```

* *Hook* [`useMemo`](/docs/hooks-faq.html#how-to-memoize-calculations) membuat jadi lebih mudah untuk mengontrol ketika ada satu-satu pembaruan turunan (*children*), mengurangi kebutuhan untuk komponen murni.

* Pada akhirnya, *Hook* [`useReducer`](/docs/hooks-reference.html#usereducer) mengurangi kebutuhan untuk mengoper *callback* secara mendalam, seperti yang dijelaskan di bawah ini.

> ### Bagaimana cara menghindari pengoperan *callback* jadi *down*? {#how-to-avoid-passing-callbacks-down}

Reviewer: ```suggestion
### Bagaimana cara menghindari pengoperan *callback* ke bawah? {#how-to-avoid-passing-callbacks-down}
```

This refers to passing callbacks down the node. Not actually putting the callback *down*.

Author: >### Bagaimana cara menghindari pengoperan *callback* ke bawah? {#how-to-avoid-passing-callbacks-down}

Suggested correction's applied.
Table of contents also updated

> Kami menemukan bahwa kebanyakan orang tidak menikmati pengoperan *callback* secara manual di sepanjang tiap tingkatan sebuah pohon komponen. Walaupun hal tersebut lebih gamblang, pengoperan *callback* secara manual bisa terasa seperti pekerjaan yang menguras tenaga.

Reviewer: ```suggestion
Kami menemukan bahwa kebanyakan orang tidak menikmati pengoperan *callback* secara manual di sepanjang tiap tingkatan sebuah diagram komponen. Walaupun hal tersebut lebih gamblang, pengoperan *callback* secara manual bisa terasa seperti pekerjaan yang menguras tenaga.
```

Author: >Kami menemukan bahwa kebanyakan orang tidak menikmati pengoperan *callback* secara manual di sepanjang tiap tingkatan sebuah diagram komponen. Walaupun hal tersebut lebih gamblang, pengoperan *callback* secara manual bisa terasa seperti pekerjaan yang menguras tenaga.

Suggested change's applied

> Dalam pohon komponen yang luas, sebuah alternatif yang kami rekomendasikan adalah untuk mengoper sebuah fungsi `dispatch` dari [`useReducer`](/docs/hooks-reference.html#usereducer) melalui *context*:

Reviewer: ```suggestion
Dalam diagram komponen yang luas, sebuah alternatif yang kami rekomendasikan adalah untuk mengoper sebuah fungsi `dispatch` dari [`useReducer`](/docs/hooks-reference.html#usereducer) melalui *context*:
```

`pohon` -> `diagram`.

Author: >Dalam diagram komponen yang luas, sebuah alternatif yang kami rekomendasikan adalah untuk mengoper sebuah fungsi `dispatch` dari [`useReducer`](/docs/hooks-reference.html#usereducer) melalui *context*:

Suggested change's applied

}
```

> Turunan apapun dalam pohon yang ada di dalam `TodosApp` bisa menggunakan fungsi `dispatch` untuk mengoper tindakan kepada `TodosApp`:

Reviewer: ```suggestion
Turunan apapun dalam diagram yang ada di dalam `TodosApp` bisa menggunakan fungsi `dispatch` untuk mengoper tindakan kepada `TodosApp`:
```

Author: >Turunan apapun dalam diagram yang ada di dalam `TodosApp` bisa menggunakan fungsi `dispatch` untuk mengoper tindakan kepada `TodosApp`:

}
```

Hal ini lebih memudahkan dari sudut pandang pemeliharaan (tidak perlu untuk terus meneruskan *callback*), dan juga menghindari masalah-masalah *callback*. Mengoper `dispatch` seperti ini merupakan rekomendasi pola untuk pembaruan-pembaruan mendalam.

Perlu diingat bahawa Anda masih bisa memilih antara mengoper *state* aplikasi sebagai *prop* (lebih gamblang) atau sebagai *context* (lebih memudahkan untuk pembaruan yang sangat mendalam). Jika Anda menggunakan *context* untuk mengoper *state* juga, gunakan dua jenis *context* berbeda -- `dispatch` context* tidak pernah berubah, jadi komponen yang membacanya tidak perlu melakukan *render* ulang kecuali juga membutuhkan *state* aplikasi.

### Bagimana cara membaca sebuah nilai yang sering berubah dari `useCallback`? {#how-to-read-an-often-changing-value-from-usecallback}

>Catatan
>

> >Kami rekomendasikan untuk [mengoper `dispatch` dalam *context*](#how-to-avoid-passing-callbacks-down) daripada satu-persatu *callback* dalam *prop*. Pendekatan di bawah hanya disebutkan di sini sebagai pelengkap dan juga cara darurat.

Reviewer: ```suggestion
>Kami rekomendasikan untuk [mengoper `dispatch` dalam *context*](#how-to-avoid-passing-callbacks-down) daripada *callback* dalam *prop* satu-persatu. Pendekatan di bawah hanya disebutkan di sini sebagai pelengkap dan juga cara darurat.
```

Author: >>Kami rekomendasikan untuk [mengoper `dispatch` dalam *context*](#how-to-avoid-passing-callbacks-down) daripada *callback* dalam *prop* satu-persatu. Pendekatan di bawah hanya disebutkan di sini sebagai pelengkap dan juga cara darurat.

Suggested change's applied

}
```

Ini bisa saja pola yang berbelit namun menunjukkan pada Anda bahwa Anda bisa melakukan optimasi darurat ini jika membutuhkannya. Bahkan lebih bisa ditangani jika Anda mengekstraknya ke dalam sebuah *custom Hook*:

```js{4,16}
function Form() {
  const [text, updateText] = useState('');

>   // Akan di memoize bahkan jika `text` berubah:

Reviewer: ```suggestion
  // Akan di-memoize bahkan jika `text` berubah:
```

Author: > // Akan di-memoize bahkan jika `text` berubah:

Suggested hyphen addition's applied

}
```

Dalam kasus-kasus tersebut, kami **tidak merekomendasikan pola ini** dan hanya untuk menunjukannya sebagai pelengkap di sini. Sebagai gantinya, akan lebih baik untuk [meghindari pengoperan *callback* lebih mendalam](#how-to-avoid-passing-callbacks-down).

## *Under the Hood* {#under-the-hood}

### Bagaimana React menghubungkan panggilan *Hook* dengan komponen? {#how-does-react-associate-hook-calls-with-components}

> React merekam jejak komponen yang baru saja mulai di-*render*. Terima kasih pada [aturan *Hooks*](/docs/hooks-rules.html), kita tahu bahwa *Hooks* hanya dipanggil dari komponen React (atau *custom Hooks* -- yang mana juga dipanggil dari komponen React).

Reviewer: ```suggestion
React merekam jejak komponen yang baru saja mulai di-*render*. Berkat [aturan *Hooks*](/docs/hooks-rules.html), kita tahu bahwa *Hooks* hanya dipanggil dari komponen React (atau *custom Hooks* -- yang mana juga dipanggil dari komponen React).
```

Author: >React merekam jejak komponen yang baru saja mulai di-*render*. Berkat [aturan *Hooks*](/docs/hooks-rules.html), kita tahu bahwa *Hooks* hanya dipanggil dari komponen React (atau *custom Hooks* -- yang mana juga dipanggil dari komponen React).

Suggested change's applied

Terdapat daftar internal dari *"memory cells"* yang berhubungkan dengan tiap komponen. Itu semua hanyalah objek-objek JavaScript dimana kita bisa menaruh data. Ketika Anda memanggil sebuah *Hook* seperti `useState()`, ia membaca *cell* terkini (atau menginisialisasikannya selama *render* pertama), dan memindahkan *pointer* ke *cell* selanjutnya. Inilah bagaimana masing-masing panggilan `useState()` mendapat *local state* sendiri-sendiri.

> ### Bagaimana *seni awal*-nya *Hooks*? {#what-is-the-prior-art-for-hooks}

Reviewer: ```suggestion
### Adakah hasil kerja sebelumnya yang menginspirasi *Hooks*? {#what-is-the-prior-art-for-hooks}
```

Author: >### Adakah hasil kerja sebelumnya yang menginspirasi *Hooks*? {#what-is-the-prior-art-for-hooks}

Suggested translation is applied.
Table of contents also updated

*Hooks* mempersatukan ide-ide dari beberapa sumber berbeda:

> * Eksperimen lama kami dengan API fungsional dalam repositori[*react-future*](https://github.com/reactjs/react-future/tree/master/07%20-%20Returning%20State).

Reviewer: ```suggestion
* Eksperimen lama kami dengan API fungsional dalam repositori [*react-future*](https://github.com/reactjs/react-future/tree/master/07%20-%20Returning%20State).
```

Missing space.

Author: >* Eksperimen lama kami dengan API fungsional dalam repositori [*react-future*](https://github.com/reactjs/react-future/tree/master/07%20-%20Returning%20State).

Added missing space

> * Eksperimen-eksperimen komunitas React dengan me-*render prop API*, termasuk [Reactions Component](https://github.com/reactions/component) dari [Ryan Florence](https://github.com/ryanflorence).

Reviewer: ```suggestion
* Eksperimen-eksperimen komunitas React dengan me-*render* API *props*, termasuk [Reactions Component](https://github.com/reactions/component) dari [Ryan Florence](https://github.com/ryanflorence).
```

Author: >* Eksperimen-eksperimen komunitas React dengan me-*render* API *props*, termasuk [Reactions Component](https://github.com/reactions/component) dari [Ryan Florence](https://github.com/ryanflorence).

Suggested change's applied

* Pengajuan [kata kunci `adopt`](https://gist.github.com/trueadm/17beb64288e30192f3aa29cad0218067) dari [Dominic Gannaway](https://github.com/trueadm) sebagai *sugar syntax* untuk me-*render* *prop*.
* *State variable* dan *state cell* dalam [DisplayScript](http://displayscript.org/introduction.html).

> * [Reducer components](https://reasonml.github.io/reason-react/docs/en/state-actions-reducer.html) dalam *ReasonReact*.

Reviewer: ```suggestion
* [Reducer components](https://reasonml.github.io/reason-react/docs/en/state-actions-reducer.html) dalam ReasonReact.
```

ReasonReact is a name of a library. No need to be italicised.

Author: >* [Reducer components](https://reasonml.github.io/reason-react/docs/en/state-actions-reducer.html) dalam ReasonReact.

* [Subscriptions](http://reactivex.io/rxjs/class/es6/Subscription.js~Subscription.html) dalam Rx.

> * [Algebraic effects](https://github.com/ocamllabs/ocaml-effects-tutorial#2-effectful-computations-in-a-pure-setting) dalam *Multicore OCaml*.

Reviewer: ```suggestion
* [Algebraic effects](https://github.com/ocamllabs/ocaml-effects-tutorial#2-effectful-computations-in-a-pure-setting) dalam *Multicore* OCaml.
```

OCaml is a language name. No need to be italicised.

Author: >* [Algebraic effects](https://github.com/ocamllabs/ocaml-effects-tutorial#2-effectful-computations-in-a-pure-setting) dalam *Multicore* OCaml.

Suggested correction is applied

> [Sebastian Markbåge](https://github.com/sebmarkbage) muncul dengan desain orisinil untuk *Hooks*, yang nantinya akan disempurnakan oleh [Andrew Clark](https://github.com/acdlite), [Sophie Alpert](https://github.com/sophiebits), [Dominic Gannaway](https://github.com/trueadm), serta anggota-anggota lain tim React.

Reviewer: ```suggestion
[Sebastian Markbåge](https://github.com/sebmarkbage) membuat desain awal dari *Hooks*, yang kemudian disempurnakan oleh [Andrew Clark](https://github.com/acdlite), [Sophie Alpert](https://github.com/sophiebits), [Dominic Gannaway](https://github.com/trueadm), serta anggota-anggota lain tim React.
```

We're talking in the past tense. It *has been* refined by Andrew Clark.

Author: >[Sebastian Markbåge](https://github.com/sebmarkbage) membuat desain awal dari *Hooks*, yang kemudian akan oleh [Andrew Clark](https://github.com/acdlite), [Sophie Alpert](https://github.com/sophiebits), [Dominic Gannaway](https://github.com/trueadm), serta anggota-anggota lain tim React.

Suggested corrections are applied
